csrc/scheduler/multi_matmul.cpp

```diff
@@ -0,0 +1,421 @@
+    cacheOperandsToSmem(bs_, bcw_smems_, params_.supported_vec_size.b);
+
+    // We add two LoadStore operators to the inputs of our fusions. The first
+    // one is for a read from global memory and the second one (below) is for a
+    // cache read. As an optimizaton, we avoid adding an operator if there's an
+    // existing LoadStoreOp present. Please note that for the second LoadStore
+    // we don't propagate the allocation domain, since the scheduler sets the
+    // allocation domain in the registers.
+    auto addSetsForCacheReads = [&](const std::vector<TensorView*>& tv_smems,
+                                    std::vector<TensorView*>& tv_rs) {
+      tv_rs.resize(tv_smems.size(), nullptr);
+      for (size_t i : c10::irange(tv_smems.size())) {
+        TensorView* tv_smem = tv_smems[i];
+        TensorView*& tv_r = tv_rs[i];
+
+        if (auto ldst = dynamic_cast<LoadStoreOp*>(tv_smem->uses().at(0))) {
+          tv_r = ldst->out()->as<TensorView>();
+          ldst->setOpType(LoadStoreOpType::LdMatrix);
+        } else {
+          tv_r = cacheAfter(
+              tv_smem,
+              LoadStoreOpType::LdMatrix,
+              CacheOp::Unspecified,
+              /*propagate_allocation_domain=*/false);
+        }
+      }
+    };
+    // Shared memory read
+    addSetsForCacheReads(acw_smems_, acrs_);
+    addSetsForCacheReads(bcw_smems_, bcrs_);
+  }
+
+  //! This calls orig->cacheAfter() and also updates the permissive graph to
+  //! reflect the new IterDomain mappings
+  TensorView* cacheAfter(
+      TensorView* orig,
+      LoadStoreOpType op_type = LoadStoreOpType::Set,
+      CacheOp cache_op = CacheOp::AllLevels,
+      bool propagate_allocation_domain = false) {
+    const std::vector<IterDomain*> orig_alloc =
+        orig->getMaybeAllocationDomain();
+
+    TensorView* c =
+        orig->cacheAfter(op_type, cache_op, propagate_allocation_domain);
+
+    if (propagate_allocation_domain) {
+      const std::vector<IterDomain*> cache_alloc =
+          c->getMaybeAllocationDomain();
+      NVF_ERROR(orig_alloc.size() == cache_alloc.size());
+      for (size_t i : c10::irange(orig_alloc.size())) {
+        ValGroup vg = graph_->toGroup(orig_alloc[i]);
+        graph_->initializeVal(cache_alloc[i], vg);
+      }
+    }
+
+    const std::vector<IterDomain*> orig_logical =
+        TensorDomain::noReductions(orig->getLogicalDomain());
+    const std::vector<IterDomain*> cache_logical = c->getLogicalDomain();
+    // in split-K we do rFactor which gives us a full = sum(partial)
+    // where partial has root domain that matches the logical domain of the
+    // original tensor. The logical domain contains Iteration transforms of the
+    // Reduction axis in the original mma output.
+    NVF_ERROR(orig_logical.size() == cache_logical.size());
+    for (size_t i : c10::irange(orig_logical.size())) {
+      ValGroup vg = graph_->toGroup(orig_logical[i]);
+      graph_->initializeVal(cache_logical[i], vg);
```

Comment:
Is this sufficient? Should we map these domains with its producer and consumer?

Reply:
Here I am just adding the ID to the `ValGroup` in the producer. Those should already be mapped appropriately with the consumer if the graph is up to date before this point I think.

Reply:
Oh, I see